tests/phpunit/tests/admin/wpUpgrader.php

```diff
@@ -0,0 +1,1430 @@
+						'files' => array(
+							'subfile1.php' => array( 'name' => 'subfile1.php' ),
+							'subfile2.php' => array( 'name' => 'subfile2.php' ),
+							'subsubdir1'   => array(
+								'files' => array(
+									'subsubfile1.php' => array(
+										'name' => 'subsubfile1.php',
+									),
+									'subsubfile2.php' => array(
+										'name' => 'subsubfile2.php',
+									),
+								),
+							),
+						),
+					),
+					'subdir2' => array(
+						'files' => array(
+							'subfile3.php' => array( 'name' => 'subfile3.php' ),
+							'subfile4.php' => array( 'name' => 'subfile4.php' ),
+							'subsubdir2'   => array(
+								'files' => array(
+									'subsubfile3.php' => array(
+										'name' => 'subsubfile3.php',
+									),
+									'subsubfile4.php' => array(
+										'name' => 'subsubfile4.php',
+									),
+								),
+							),
+						),
+					),
+				),
+				'path'         => 'custom_path/',
+			),
+		);
+	}
+
+	/**
+	 * Tests that `WP_Upgrader::clear_destination()` returns early with `true`
+	 * when the destination does not exist.
+	 *
+	 * @ticket 54245
+	 *
+	 * @covers WP_Upgrader::clear_destination
+	 */
+	public function test_clear_destination_should_return_early_when_the_destination_does_not_exist() {
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'is_writable' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'chmod' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'delete' );
+
+		$destination = DIR_TESTDATA . '/upgrade/';
+
+		$this->wp_filesystem_mock
+				->expects( $this->once() )
+				->method( 'dirlist' )
+				->with( $destination )
+				->willReturn( false );
+
+		$this->assertTrue( $this->instance->clear_destination( $destination ) );
+	}
+
+	/**
+	 * Tests that `WP_Upgrader::clear_destination()` clears
+	 * the destination directory.
+	 *
+	 * @ticket 54245
+	 *
+	 * @covers WP_Upgrader::clear_destination
+	 */
+	public function test_should_clear_the_destination_directory() {
+		$destination = DIR_TESTDATA . '/upgrade/';
+
+		$this->wp_filesystem_mock
+				->expects( $this->once() )
+				->method( 'dirlist' )
+				->with( $destination )
+				->willReturn( array() );
+
+		$this->wp_filesystem_mock
+				->expects( $this->once() )
+				->method( 'delete' )
+				->with( $destination )
+				->willReturn( true );
+
+		$this->assertTrue( $this->instance->clear_destination( $destination ) );
+	}
+
+	/**
+	 * Tests that `WP_Upgrader::clear_destination()` returns a WP_Error object
+	 * if files are not writable.
+	 *
+	 * @ticket 54245
+	 *
+	 * @covers WP_Upgrader::clear_destination
+	 */
+	public function test_should_return_wp_error_if_files_are_not_writable() {
+		$this->instance->generic_strings();
+
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'delete' );
+
+		if ( ! defined( 'FS_CHMOD_DIR' ) ) {
+			define( 'FS_CHMOD_DIR', 0755 );
+		}
+
+		if ( ! defined( 'FS_CHMOD_FILE' ) ) {
+			define( 'FS_CHMOD_FILE', 0644 );
+		}
```

Comment:
Why are these defines needed here ?
If these constants are needed for the application to work, they should be defined in WP `src`.
Defining constants cannot be undone, so you are now changing the test environment for all tests which will run _after_ this test.

Reply:
If these are PHP native constants, they should probably be defined in the `wp-includes/compat.php` file, not in the `WP_Upgrader` class.
And IIRC, that file is loaded in the test bootstrap.

More interestingly: how come the tests fail, but no errors are thrown IRL ? Are they already declared somewhere in WP Core, just not in a file which is loaded when running these tests ?

Reply:
They're declared in [WP_Filesystem()](https://github.com/WordPress/wordpress-develop/blob/6.0/src/wp-admin/includes/file.php#L2024-L2030), which is not used in the unit tests.